File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
from itertools import permutations as per

def solution(expression):
    
    answer = 0
    for order in per("+-*", 3):
        answer = max(abs(int(calc(expression, "".join(order)))), answer)
    return answer


def calc(ex, symbols):
    if not symbols:
        return ex
    
    answer = symbols[0].join(map(lambda x : calc(x, symbols[1:]), ex.split(symbols[0])))
    return str(eval(answer))
        

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public long solution(String expression) {
        long answer = 0;
        
        for (String symbols : List.of("+-*", "+*-", "-+*", "-*+", "*+-", "*-+")) {
            answer = (long) Math.max(Math.abs(calc(expression, symbols, 0)), answer);
        }
        
        return answer;
    }
    
    public long calc(String ex, String symbols, int index) {
        if (index == 3) {
            return Long.parseLong(ex);
        }
        
        String symbol = String.valueOf(symbols.charAt(index));
            
        if (symbol.equals("+")) {
            return Arrays.stream(ex.split("\\" + symbol))
            .mapToLong(x -> calc(x, symbols, index + 1))
            .reduce((a, b) -> a + b).orElse(0);
        }

        if (symbol.equals("-")) {
    
            return Arrays.stream(ex.split(symbol))
            .mapToLong(x -> calc(x, symbols, index + 1))
            .reduce((a, b) -> a - b).orElse(0);
        }
        
        return Arrays.stream(ex.split("\\" + symbol))
        .mapToLong(x -> calc(x, symbols, index + 1))
        .reduce((a, b) -> a * b).orElse(0);
    }
}